### PR TITLE
Decouple LTIHService from LTILaunchResource.provisioning_enabled

### PR DESCRIPTION
**Pull request by seanh**
Simplify `ApplicationInstanceGetter` and decouple `LTIHService` from `LTILaunchResource.provisioning_enabled`. This is part of decoupling `LTIHService` from `LTILaunchResource` entirely, so that `LTIHService` can be used in other request contexts (specifically: so that it can be used in new sync API requests).

* Change `ApplicationInstanceGetter` to receive `request.params["oauth_consumer_key"]` from pyramid_services rather than requiring it to be passed in to each method. This changes `ApplicationInstanceGetter.foo(consumer_key)` to just `ApplicationInstanceGetter.foo()` simplifying code that uses it

* Remove `LTILaunchResource.provisioning_enabled` and `LTILaunchResource.lms_url`. These were just wrappers for `ApplicationInstanceGetter` that injected `request.params["oauth_consumer_key"]`. Since `ApplicationInstanceGetter`'s factory now injects the consumer key, these no longer add anything. Change using code to use `ApplicationInstanceGetter` directly instead. This reduces code such as `LTIHService`'s dependency on `LTILaunchResource` which moves it closer to being usable in other request contexts.

We could go one step further and expose the request's `ApplicationInstance` model itself directly, perhaps as `request.application_instance`. But this is far enough for what I need to achieve right now, which is just to decouple `LTIHService` from `LTILaunchResource`

### CODE REVIEW

**Comment by seanh** [comment]
It can still be reviewed but I'm slapping **do not merge** on this 'cos manual testing shows that it might have broken something

File: tests/unit/lms/services/application_instance_getter_test.py [hunk 1]
```diff
@@ -12,18 +12,19 @@
...
 
     def test_developer_key_returns_None_if_ApplicationInstance_has_no_developer_key(
         self, ai_getter, test_application_instance
     ):
         test_application_instance.developer_key = None
 
-        assert ai_getter.developer_key("TEST_CONSUMER_KEY") is None
+        assert ai_getter.developer_key() is None
 
+    @pytest.mark.usefixtures("unknown_consumer_key")
```
Comment (jon-betts) [inline tests/unit/lms/services/application_instance_getter_test.py:24]: I like this pattern. Prevents the "we don't use this fixture, did we import it by mistake?" question

**Comment by seanh** [comment]
The bug I found in manual testing was that creating and launching Canvas file assignments was broken. It turns out the `ApplicationInstanceGetter` was being used during API requests that're made during Canvas files launches. During an API request `request.params.get("oauth_consumer_key")` is `None` so it was failing to ever find the user's Canvas access token in the DB.

The fix is to read `request.lti_user.oauth_consumer_key` instead. That is set by the authentication machinery during both launch requests and API requests.

Unfortunately reading `request.lti_user` indirectly calls `LaunchVerifier` and `LaunchVerifier` uses [`ApplicationInstanceGetter`. So if `ApplicationInstanceGetter` (or its factory) reads `request.lti_user` you get a circular dependency.

So:

1. Make `LaunchVerifier` not depend on `ApplicationInstanceGetter`, so you won't get a circular dependency: https://github.com/hypothesis/lms/pull/1637/commits/5075815292adfa6e08c2c450d337a838c12cf6a9

2. Now fix the bug by changing `ApplicationInstanceGetter`'s factory to read  `request.lti_user.oauth_consumer_key`: https://github.com/hypothesis/lms/pull/1637/commits/78306ace25b2bb5b8089b80e9d22d7a2c0ee17f3

This can be reviewed again now but I'm leaving it **do not merge** as I want to more thoroughly test it before merging it.

**Comment by seanh** [comment]
Alright, I've tested this to my satisfaction. Removing **do not merge**

**Comment by seanh** [comment]
> I think moving to the query to the db object makes sense here, but raises an inevitable question about where service layer and db layer stuff separates. I think the purist way to do this would be to have a separate pure "ApplicationInstanceGetter" service that both depend on that does this, but honestly who's got time for that.

I also think that simple single-model getter methods and the like are probably better on the models themselves. Although it's annoying that you have to pass the DB session into them.

Some guidelines for this:

* One of the reasons for wanting to keep the models "thin" is that models tend to get passed around throughout the whole codebase so having a complex model with lots of logic and behaviour couples the whole codebase together. This doesn't have to mean that models don't have _any_ methods, though.

* Logic that deals with multiple different model classes doesn't belong in the models. It quickly gets complicated and can lead to circular imports. Move that up into services.

* Methods that need external services don't belong in the models. Move those up into services.

* We don't want web framework stuff like the request or ACLs or URL-generation or HTML or JSON etc etc in the models.

* So a model class can contain logic and behaviour as long as that logic and behaviour pertains only to that model class and not to other things, such as other model classes, the web layer, or external services

* A simple single-model method that finds all the data it needs in that one model, belongs on that model class.

The original ADR that was written when the service layer was introduced (<https://github.com/hypothesis/h/blob/master/docs/arch/adr-002.rst>) and the blog post that it links to (<http://dev.nando.audio/2014/04/01/large_apps_with_sqlalchemy__architecture.html>) are both pretty good.

So I think the `get()` method on `models.ApplicationInstance` is correct. But I do also think that the `ApplicationInstanceGetter` service, as much as it may not be correctly designed currently, does add something that belongs above the model:

1. It gets the particular `ApplicationInstance` for this request. You wouldn't want that logic in the models 'cos then you'd have the `request` in the models

2. It encapsulates the model. Views don't deal with `models.ApplicationInstance` objects. They just deal with consumer keys and secrets, developer keys and secrets, etc as strings. Not entirely sure whether this second thing is necessary or a good idea. ([The blog post](http://dev.nando.audio/2014/04/01/large_apps_with_sqlalchemy__architecture.html) says "models can be returned by each layer, all the way to the template at the end of a request" and I think that may make sense.)

File: tests/unit/lms/services/launch_verifier_test.py [hunk 2]
```diff
@@ -206,4 +210,8 @@ def sign(pyramid_request):
     )
 
 
-pytestmark = pytest.mark.usefixtures("ai_getter")
+@pytest.fixture(autouse=True)
+def models(patch):
+    models = patch("lms.services.launch_verifier.models")
+    models.ApplicationInstance.get.return_value = mock.Mock(shared_secret="TEST_SECRET")
```
[the remark this answers is not in the document]
Comment (seanh) [reply tests/unit/lms/services/launch_verifier_test.py:216]: I think it might be nicer to use a mock here for better test isolation. The `LaunchVerifier` service doesn't really depend on `models.ApplicationInstance` as such. It really just depends on the classmethod `models.ApplicationInstance.get()` and its contract, and then on the `.shared_secret` attribute of the returned `ApplicationInstance`. That said, I forgot to make it a `create_autospec()`.

A possible downside is that mocking might be slower here

File: lms/services/application_instance_getter.py [hunk 3]
```diff
@@ -50,75 +46,66 @@ def developer_secret(self, consumer_key):
...
-        except NoResultFound as err:
-            raise ConsumerKeyError() from err
+        application_instance = models.ApplicationInstance.get(
+            self._db, self._consumer_key
+        )
+
+        if application_instance is None:
+            raise ConsumerKeyError()
+
+        return application_instance
```
[the remark this answers is not in the document]
Comment (seanh) [reply lms/services/application_instance_getter.py:103]: Yeah that seems fine.

If `ConsumerKeyError` is moved from a services error to a models error, then it might be a layering violation for code-above-the-services (i.e. views and resources) to be catching a models error, rather than the services encapsulating the models so that code-above-services only catches services errors.

On the other hand code-above-services is also above models, so maybe there is not layering violation in code-above-services using the models (as long as it's just simple things like catching a models exception or calling a models method, not something like doing a complex SQLAlchemy query).

The only above-services thing that uses `ConsumerKeyError` anyway is `JSConfig` in `resources`.

**Comment by seanh** [comment]
> Not mad keen on get() as a function name for two reasons. It's a pain to grep for, and usually a plain find() or get() method on a db object will accept the primary id. I looked through the other classes and there's not really a pattern in place. I'd suggest get_by_consumer_key to allow space for future queries of another kind.

There is `db.query(SomeModel).get(primary_key_value)`, which is a SQLAlchemy-provided method _on the query object_ rather than on the model classes.

I've renamed `ApplicationInstance.get()` to `ApplicationInstance.get_by_consumer_key()`, that makes sense as a naming pattern for these

File: lms/services/launch_verifier.py [hunk 4]
```diff
@@ -54,13 +55,15 @@ def _verify(self):
...
-        try:
-            shared_secret = self._request.find_service(name="ai_getter").shared_secret()
-        except ConsumerKeyError:  # pylint: disable=try-except-raise
-            raise
+        application_instance = models.ApplicationInstance.get(
+            self._request.db, consumer_key
+        )
+
+        if not application_instance:
+            raise ConsumerKeyError()
```
[the remark this answers is not in the document]
Comment (seanh) [reply lms/services/launch_verifier.py:63]: Done

File: tests/unit/lms/models/test_application_instance.py [hunk 5]
```diff
@@ -61,6 +61,22 @@ def test_requesters_email_cant_be_null(self, db_session, application_instance):
         with pytest.raises(IntegrityError, match="requesters_email"):
```
[the remark this answers is not in the document]
Comment (seanh) [reply tests/unit/lms/models/test_application_instance.py:61]: It should be for consistency but it is actually being run all the time

File: lms/services/application_instance_getter.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (seanh) [reply lms/services/application_instance_getter.py:103]: Done

[the remark this answers is not in the document]
Comment (seanh) [reply lms/services/application_instance_getter.py]: I decided not to do this. It turns out that `services.ConsumerKeyError` is a subset of `services.LTILaunchVerificationError` and that fact is made use of to change behaviour in a number of places.

If  `services.ConsumerKeyError(LTILaunchVerificationError)` becomes `models.ConsumerKeyError` and therefore no longer a `services.LTILaunchVerificationError` subclass, stuff breaks.

You'll end up having to keep catch `models.ConsumerKeyError` and raise `services.ConsumerKeyError(LTILaunchVerificationError)`, which removes the simplification and de-duplication benefit, and you'd have to keep a `services.ConsumerKeyError(LTILaunchVerificationError)` class around so now you have two exception classes.

Basically `ConsumerKeyError` is semantically absolutely a services-layer error and that's important

**Comment by seanh** [comment]
Rebased
